### PR TITLE
GH-2753: Customize Fuseki main arguments

### LINKED ISSUE #2753: Improve ability to support custom arguments for Fuseki Servers
### Version

5.1.0

### Feature

The Fuseki Modules mechanism makes it possible to add extensions to Fuseki but modules have to rely upon external configuration sources e.g. System Properties, Environment Variables, Fuseki Configuration file etc. because there isn't a meaningful way to pass configuration via the command line.

`FusekiMain` is built on Jena's general CLI framework which means we can in principal extend `FusekiMain` and add additional arguments via the constructor, or use the static `FusekiMain.addArgModule()` method to register global extensions.

However, it's hard to do anything meaningful with these custom arguments, `processModulesAndArgs()` currently populates a private `ServerConfig` class which is then used to actually populate a `FusekiServer.builder()` via a private static `applyServerArgs()` which cannot be modified.  Thus even if a custom argument is introduced the only way for it to pass configuration to a module would be by placing it into System Properties, static variable or another JVM wide store that the module can later read back.

As a practical example of this we're looking to introduce support for specifying Fuseki configuration via a YAML configuration file for which we'd like to add a `--yaml-config <yaml-file>` argument and then have a `FusekiModule` do a `prepare()` step which reads in the `<yaml-file>` and modifies the `FusekiServer.Builder` accordingly.  Right now we can't do that cleanly because there's no way to get the value of `<yaml-file>` passed into the builder for later access without resorting to passing it out-of-band.

It would be better if we were able to carry extra configuration information via `ServerConfig` and override `applyServerArgs()` so that we can apply that to the builder this would offer users who want to extend `FusekiMain` much more flexibility.

### Are you interested in contributing a solution yourself?

Yes

### PR DESCRIPTION
**Pull request by afs**
GitHub issue resolved #2753

----

 - [x] Tests are included.
 - [ ] Documentation change and updates are provided for the [Apache Jena website](https://github.com/apache/jena-site/)
 - [ ] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/cmds/DSGSetup.java [hunk 1]
```diff
@@ -46,40 +61,78 @@
...
+        serverArgs.datasetDescription = "in-memory";
+        serverArgs.dsg = DatasetGraphFactory.createTxnMem();
+        serverArgs.allowUpdate = true;
+    }
+
+    /*package*/ static void setupFile(Logger log, List<String> filenames, ServerArgs serverArgs) {
+        serverArgs.datasetDescription = "in-memory, with files loaded";
+        serverArgs.dsg = DatasetGraphFactory.createTxnMem();
+
+        for(String filename : filenames ) {
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/cmds/DSGSetup.java:110]: s/for(/for (/ ?

**Review by kinow — APPROVED** [review]
Skimmed through, found nothing that seems wrong/broken. :+1: :clap:

File: jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/sys/FusekiModules.java [hunk 2]
```diff
@@ -23,16 +23,51 @@
...
  * @see FusekiAutoModules
  */
 public class FusekiModules {
 
+    private static FusekiModules autoLoadedFusekiModules = FusekiAutoModules.get();
+    // Never null, maybe empty
+    private static FusekiModules systemFusekiModules = autoLoadedFusekiModules;
+
+    /**
+     * There is a system wide set of modules used when o other are indicated.
```
Comment (rvesse) [inline jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/sys/FusekiModules.java:40]:
```suggestion
     * There is a system wide set of modules used when no other are indicated.
```

**Review by rvesse — APPROVED** [review]
Looks good to me, and looking at the test cases this covers all the scenarios that motivated my original ticket #2753

File: jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/cmds/DSGSetup.java (hunk 1 above)
Comment (afs) [reply jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/cmds/DSGSetup.java:110]: Fixed.

**Comment by afs** [comment]
> Looks good to me, and looking at the test cases this covers all the scenarios that motivated my original ticket #2753

There a few niggles in the command argument processing that could be improved and the overall system assumes a degree of cooperation between elements not to tread on each other. These are best dealt with from experience.

Please do let us know how using it goes.

**Comment by afs** [comment]
Apologies - I forgot to push the fixes mentioned above. I'll include them in an args clearup PR.

**Pull request merged**